src/core/sock/sockinfo_tcp.cpp

```diff
@@ -894,80 +889,163 @@ ssize_t sockinfo_tcp::tx(xlio_tx_call_attr_t &tx_arg)
     return m_ops->tx(tx_arg);
 }
 
-static inline bool cannot_do_requested_partial_write(const tcp_pcb &pcb,
+static inline bool cannot_do_requested_partial_write(size_t sndbuf_available,
                                                      const xlio_tx_call_attr_t &tx_arg,
-                                                     bool is_blocking, size_t total_iov_len)
+                                                     size_t total_iov_len)
 {
-    return !BLOCK_THIS_RUN(is_blocking, tx_arg.attr.flags) &&
-        (tx_arg.xlio_flags & TX_FLAG_NO_PARTIAL_WRITE) &&
-        unlikely(tcp_sndbuf(&pcb) < total_iov_len);
+    return (tx_arg.xlio_flags & TX_FLAG_NO_PARTIAL_WRITE) &&
+        unlikely(sndbuf_available < total_iov_len);
 }
 
-static inline bool tcp_wnd_unavalable(const tcp_pcb &pcb, size_t total_iov_len)
-{
 #ifdef DEFINED_TCP_TX_WND_AVAILABILITY
-    return !tcp_is_wnd_available(&pcb, total_iov_len);
+#define TCP_WND_UNAVALABLE(pcb, total_iov_len) !tcp_is_wnd_available(&pcb, total_iov_len)
 #else
-    NOT_IN_USE(pcb);
-    NOT_IN_USE(total_iov_len);
-    return false;
+#define TCP_WND_UNAVALABLE(pcb, total_iov_len) false
 #endif
+
+static inline bool is_invalid_iovec(const iovec *iov, size_t sz_iov)
+{
+    return iov == nullptr || sz_iov == 0;
 }
 
 ssize_t sockinfo_tcp::tcp_tx(xlio_tx_call_attr_t &tx_arg)
```

Comment:
We have now 3 pathes. tcp_tx tcp_tx_slow_path and tcp_tx_express. When tcp_tx is ZC. don't we want to utilize tcp_tx_express for tcp_tx? why must we have 3 pathes?

Reply:
Adding appropriate comments

Reply:
ok